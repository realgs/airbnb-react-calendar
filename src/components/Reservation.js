import React from 'react';
import Calendar from './Calendar';
import Header from './Header';
import Costs from './Costs';
import { stayDetails } from '../data/stayDetails';
import { round2Decimals } from '../helpers/calendar';

export default class Reservation extends React.Component {
  constructor(props) {
    super(props);
    this.setStayLength = this.setStayLength.bind(this);
    this.handleReservation = this.handleReservation.bind(this);
    this.state ={
      user: undefined,
      price: 0,
      currency: '$',
      rating : 0,
      reviews : 0,
      bonus : 0,
      cleaningFee : 0,
      stayLength : 0,
      minStay: undefined,
      total: 0,
      serviceFee: 0,
      bonusUsed: 0,
      amount: 0,
      serviceFee: 0,
      total: 0,
    };
  }
  componentDidMount(){
    this.setState({
      ...stayDetails
    });
  }
  componentDidUpdate(){
  }

  setStayLength (stayLength) {
    this.setState({ stayLength });
    const amount = round2Decimals(this.state.price * stayLength);
    const serviceFee = round2Decimals(this.state.price * stayLength * 0.1);
    const total = round2Decimals(amount + this.state.cleaningFee + serviceFee - this.state.bonus);
    this.setState({ amount, serviceFee, total });
  }

  handleReservation() {
    if (this.state.user){
      if (this.state.stayLength != 0) {
        const reservation = {
          user: this.state.user,
          amount: this.state.amount,
          serviceFee: this.state.serviceFee,
          total: this.state.total,
          stayLength: this.state.stayLength,
          discount: this.state.total > 0 ? this.state.bonus : 0
        };
        console.log(reservation);
      } else {

      }
    }
  }
  render(){
    return (
      <div className="wrapper">
        <Header
          price={this.state.price}
          currency={this.state.currency}
          rating={this.state.rating}
          reviews={this.state.reviews}
        />
        <Calendar
          minStay={this.state.minStay}
          setStayLength={this.setStayLength}
        />
        <Costs
          price={this.state.price}
          currency={this.state.currency}
          stayLength={this.state.stayLength}
          cleaningFee={this.state.cleaningFee}
          bonus={this.state.bonus}
          amount={this.state.amount}
          serviceFee={this.state.serviceFee}
          total={this.state.total}
        />
        <button className="button__reservation" onClick={this.handleReservation}>
          Request to Book
        </button>
      </div>
    );
  }
}
