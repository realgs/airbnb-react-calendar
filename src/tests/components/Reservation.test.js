import React from 'react';
import { shallow } from 'enzyme';
import Reservation from '../../components/Reservation';

test('Should render with datepicker hidden by default.', () => {
  const wrapper = shallow(<Reservation />);
  const calendarWrapper = wrapper.find('Calendar').dive();
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.state('dateMissing')).toBe(false);
  expect(wrapper.find('Calendar').exists()).toBe(true);
  expect(wrapper.find('Calendar').prop('dateMissing')).toBe(false);
  expect(calendarWrapper.state('hidePicker')).toBe(true);
});

test('Should ask for dates of reservation after submitting with no data provided.', () => {
  const wrapper = shallow(<Reservation />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('.button__reservation').simulate('click', {
    preventDefault: () => { }
  });
  expect(wrapper.state('dateMissing')).toBe(true);
  expect(wrapper.find('Calendar').exists()).toBeTruthy();
  expect(wrapper.find('Calendar').prop('dateMissing')).toBe(true);

  const calendarWrapper = wrapper.find('Calendar').dive();
  const calendarInstance = calendarWrapper.instance();
  calendarInstance.componentDidMount();
  calendarInstance.componentDidUpdate();
  expect(calendarWrapper.state('hidePicker')).toBe(false);
  expect(calendarWrapper).toMatchSnapshot();
  expect(wrapper).toMatchSnapshot();
});

test('Should set stay length correctly.', () => {
  const stayLength = 10;
  const wrapper = shallow(<Reservation />);
  wrapper.find('Calendar').prop('setStayLength')(stayLength);
  expect(wrapper.state('stayLength')).toEqual(Math.abs(stayLength));
});

test('Should set stay length correctly for negative values.', () => {
  const stayLength = -10;
  const wrapper = shallow(<Reservation />);
  wrapper.find('Calendar').prop('setStayLength')(stayLength);
  expect(wrapper.state('stayLength')).toEqual(Math.abs(stayLength));
});

test('Should set stay length correctly for no stay.', () => {
  const stayLength = 0;
  const wrapper = shallow(<Reservation />);
  wrapper.find('Calendar').prop('setStayLength')(stayLength);
  expect(wrapper.state('stayLength')).toEqual(Math.abs(stayLength));
});

test('Should set stay length correctly for null.', () => {
  const stayLength = null;
  const wrapper = shallow(<Reservation />);
  wrapper.find('Calendar').prop('setStayLength')(stayLength);
  expect(wrapper.state('stayLength')).toEqual(0);
});

test('Should set stay length correctly for undefined.', () => {
  const stayLength = undefined;
  const wrapper = shallow(<Reservation />);
  wrapper.find('Calendar').prop('setStayLength')(stayLength);
  expect(wrapper.state('stayLength')).toEqual(0);
});

test('dateNotMissing() should toggle dateMissing back to false', () => {
  const wrapper = shallow(<Reservation />);
  wrapper.find('.button__reservation').simulate('click', {
    preventDefault: () => { }
  });
  expect(wrapper.state('dateMissing')).toBe(true);
  wrapper.find('Calendar').prop('dateNotMissing')();
  expect(wrapper.state('dateMissing')).toBe(false);
});
