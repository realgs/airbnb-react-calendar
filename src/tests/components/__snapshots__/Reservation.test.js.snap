// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should ask for dates of reservation after submitting with no data provided. 1`] = `
<div
  className="wrapper"
>
  <Header
    currency="$"
    price={19.9}
    rating={3.5}
    reviews={123}
  />
  <Calendar
    dateMissing={false}
    dateNotMissing={[Function]}
    minStay={2}
    setStayLength={[Function]}
  />
  <Costs
    amount={0}
    bonus={10}
    cleaningFee={10}
    currency="$"
    price={19.9}
    serviceFee={0}
    stayLength={0}
    total={0}
  />
  <button
    className="button__reservation"
    onClick={[Function]}
  >
    Request to Book
  </button>
</div>
`;

exports[`Should ask for dates of reservation after submitting with no data provided. 2`] = `
<div
  className="calendar"
>
  <div
    className="calendar__header"
  >
    <span>
      Dates
    </span>
    <div
      className="calendar__dates"
    >
      <button
        className="calendar__dates__checkin calendar__dates__unassigned"
        id="checkin"
        onClick={[Function]}
      >
        Check In
      </button>
      <img
        alt="date from - to separator"
        className="calendar__dates__separator"
        src="./images/right-arrow.png"
      />
      <button
        className="calendar__dates__checkout calendar__dates__unassigned"
        id="checkout"
        onClick={[Function]}
      >
        Check Out
      </button>
    </div>
  </div>
  <div
    className="calendar__picker hidden"
    id="datePicker"
  >
    <svg
      className="calendar__bubble"
      focusable="false"
      role="presentation"
    >
      <path
        className="_whdw9f"
        d="M0,10 20,10 10,0z"
      />
      <path
        className="_c3dsty"
        d="M0,10 10,0 20,10"
      />
    </svg>
    <div
      className="calendar__nav"
    >
      <button
        className="calendar__arrow__button"
        onClick={[Function]}
      >
        <img
          alt="previous month"
          className="calendar__arrow__left"
          src="./images/left-arrow.png"
        />
      </button>
      <div>
        January 2019
      </div>
      <button
        className="calendar__arrow__button"
        onClick={[Function]}
      >
        <img
          alt="next month"
          className="calendar__arrow__right"
          src="./images/right-arrow.png"
        />
      </button>
    </div>
    <Month
      candidate={null}
      current={
        Object {
          "month": 0,
          "year": 2019,
        }
      }
      from={null}
      handleDeleteCandidate={[Function]}
      handleSetCandidate={[Function]}
      handleSetDay={[Function]}
      minStay={2}
      naDays={Array []}
      setSecondDate={false}
      to={null}
      unavailable={
        Array [
          "2018-11-01",
          "2018-11-02",
          "2018-11-14",
          "2018-11-16",
          "2018-11-20",
          "2018-11-25",
          "2018-11-26",
          "2018-12-01",
          "2018-12-02",
          "2018-12-03",
          "2018-12-15",
          "2018-12-16",
          "2018-12-17",
          "2019-01-05",
          "2019-01-06",
          "2019-01-07",
          "2019-01-08",
          "2019-01-20",
          "2019-01-21",
          "2019-01-22",
          "2019-01-26",
          "2019-02-15",
          "2019-02-16",
          "2019-02-17",
        ]
      }
    />
    <SubscriptInfo
      clearDates={[Function]}
      lastModified={null}
      lastUpdate={2}
    />
  </div>
</div>
`;

exports[`Should ask for dates of reservation after submitting with no data provided. 3`] = `
<div
  className="wrapper"
>
  <Header
    currency="$"
    price={19.9}
    rating={3.5}
    reviews={123}
  />
  <Calendar
    dateMissing={true}
    dateNotMissing={[Function]}
    minStay={2}
    setStayLength={[Function]}
  />
  <Costs
    amount={0}
    bonus={10}
    cleaningFee={10}
    currency="$"
    price={19.9}
    serviceFee={0}
    stayLength={0}
    total={0}
  />
  <button
    className="button__reservation"
    onClick={[Function]}
  >
    Request to Book
  </button>
</div>
`;

exports[`Should render with datepicker hidden by default. 1`] = `
<div
  className="wrapper"
>
  <Header
    currency="$"
    price={19.9}
    rating={3.5}
    reviews={123}
  />
  <Calendar
    dateMissing={false}
    dateNotMissing={[Function]}
    minStay={2}
    setStayLength={[Function]}
  />
  <Costs
    amount={0}
    bonus={10}
    cleaningFee={10}
    currency="$"
    price={19.9}
    serviceFee={0}
    stayLength={0}
    total={0}
  />
  <button
    className="button__reservation"
    onClick={[Function]}
  >
    Request to Book
  </button>
</div>
`;
